fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * AgilityContextTest.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.directory;
+
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.record.lucene.LuceneRecordContextProperties;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStoreTestBase;
+import com.apple.foundationdb.record.provider.foundationdb.properties.RecordLayerPropertyStorage;
+import com.apple.foundationdb.tuple.Tuple;
+import com.apple.test.Tags;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.util.concurrent.ExecutionException;
+import java.util.stream.IntStream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Tests for AgilityContext.
+ */
+@Tag(Tags.RequiresFDB)
+class AgilityContextTest extends FDBRecordStoreTestBase {
+    int loopCount = 20;
+    int threadCount = 5; // if exceeds a certain size, may cause an execution pool deadlock
```

Comment:
The ForkJoinPool deadlocks is indeed very concerting, and - FMHO - should be addressed to avoid future issues.
In the context of this test, we have multiple threads that are launched together and do very little in an identical flow, therefore have a **much** higher probability of colliding in that deadlock than the 10 (or more) threads that are triggered by the `indexWriter`. Hence the extra restriction. 
Given enough time, if this deadlock will actually ever happen (let's say - once every 10 years), the merge will be frozen for a while (10 minutes? I'm not sure) and then will get a "no activity" exception (or at least that what've happen in my local environment when it was tested).